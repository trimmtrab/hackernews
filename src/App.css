.page {
  /* margin: 20px; */
}

.interactions {
  text-align: center;
  height: 50px;
}

.content {
  position: absolute;
  text-align: center;
  top: 50px;
  left: 10px;
  right: 10px;
  bottom: 0px;
  overflow-y: scroll;
}

.table-header {
  display: flex;
  line-height: 24px;
  font-size: 16px;
  padding: 0 10px;
  justify-content: space-between;
}

.table-empty {
  margin: 200px;
  text-align: center;
  font-size: 16px;
}

.table-header > span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 5px;
}
