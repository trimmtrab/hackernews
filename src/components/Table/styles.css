.table {
  margin: 20px 0;
  margin-top: 30px;
}

.table-row {
  display: flex;
  line-height: 24px;
  white-space: nowrap;
  margin: 10px 0;
  padding: 10px;
  background: #ffffff;
  border: 1px solid #e3e3e3;
}

.table-row > span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 5px;
}
