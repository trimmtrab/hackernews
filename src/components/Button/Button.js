import React from 'react';
import PropTypes from 'prop-types';

import './styles.css';

const Button = ({ onClick, className, children }) => {
  return (
    <button
      onClick={onClick}
      className={className}
      type="button"
    >
      {children}
    </button>
  );
}

Button.defaultProps = {
  className: '',
}

Button.propTypes = {
  // without isRequired can also be null or undefined
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
}

export default Button;
