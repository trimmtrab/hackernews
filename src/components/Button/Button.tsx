import React from 'react';

import './styles.css';

type Props = {
  children?: JSX.Element | JSX.Element[] | string,
  className?: string,
  onClick: () => void,
  showSpinner?: boolean,
};

const Button = ({
  children,
  className, 
  onClick = () => {},
  showSpinner = false,
}: Props) => {
  return (
    <button
      className={className}
      onClick={onClick}
      type="button"
    >
      {
        showSpinner ?
        '...': // TODO: make it actual spinner
        children
      }
    </button>
  );
}

export default Button;
