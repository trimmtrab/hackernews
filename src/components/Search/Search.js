import React from 'react';
import PropTypes from 'prop-types';

import './styles.css'

const Search = ({ 
  children,
  onChange, 
  onSubmit, 
  value,
}) => (
  <form onSubmit={onSubmit} className="search">
    <input
      type="text"
      value={value}
      onChange={onChange}
    />
    <button type="submit">
      {children}
    </button>
  </form>
);

Search.defaultProps = {
  onChange: () => {},
  onSubmit: () => {},
}

Search.propTypes = {
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  onSubmit: PropTypes.func,
  children: PropTypes.node,
}

export default Search;
