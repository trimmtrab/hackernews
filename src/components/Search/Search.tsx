import React from 'react';

import './styles.css'

type Props = {
  children?: JSX.Element | JSX.Element[] | string,
  onChange?: (event: React.FormEvent<HTMLInputElement>) => void,
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void,
  value?: string,
};

const Search = ({ 
  children,
  onChange = () => {}, 
  onSubmit = () => {}, 
  value = '',
}: Props) => (
  <form 
    className="search"
    onSubmit={onSubmit} 
  >
    <input
      onChange={onChange}
      type="text"
      value={value}
    />
    <button type="submit">
      {children}
    </button>
  </form>
);

export default Search;
