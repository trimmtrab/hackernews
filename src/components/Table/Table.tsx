import React from 'react';

import Button from '../Button';

import './styles.css';

type Props = {
  list: {
    author?: string,
    num_comments?: number,
    objectID: string,
    points?: number,
    title?: string,
    url?: string,
  }[],
  onDismiss: (id: any) => void,
};

const Table = ({ 
  list, 
  onDismiss 
}: Props) => {
  const largeColumn = {
    width: '40%',
  };

  const midColumn = {
    width: '30%',
  };

  const smallColumn = {
    width: '10%',
  };

  return (
    <div className="table">
      {list.map(item =>
        <div key={item.objectID} className="table-row">
          <span style={largeColumn}>
            <a href={item.url}>{item.title}</a>
          </span>
          <span style={midColumn}>
            {item.author}
          </span>
          <span style={smallColumn}>
            {item.num_comments}
          </span>
          <span style={smallColumn}>
            {item.points}
          </span>
          <span style={smallColumn}>
            <Button
              className="button-inline"
              onClick={() => onDismiss(item.objectID)}
            >
              Dismiss
            </Button>
          </span>
        </div>
      )}
    </div>
  );
}

export default Table;
