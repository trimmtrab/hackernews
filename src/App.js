import React, { Component } from 'react';
import axios from 'axios';

import Button from './components/Button';
import Search from './components/Search';
import Table from './components/Table';

import './App.css';

const DEFAULT_QUERY = 'Redux';
const DEFAULT_HPP = 100;

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      results: null,
      // updates only when search form is submitted
      searchKey: '',
      // updates according to current input value
      searchTerm: DEFAULT_QUERY,
      // indicates error during fetch
      error: null,
      fetchingStories: false,
    };
  }

  componentDidMount() {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm);
  }

  fetchSearchTopStories = (searchTerm, page = 0) => {
    this.setState({ fetchingStories: true });
    axios(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(result => this.setSearchTopStories(result))
      .catch(error => this.setState({ error }))
      .then(() => { this.setState({ fetchingStories: false }); });
  }

  // client makes a request to the API only once
  needsToSearchTopStories = searchTerm => !this.state.results[searchTerm]; // eslint-disable-line

  onDismiss = (id) => {
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const isNotId = item => item.objectID !== id;
    const updatedHits = hits.filter(isNotId);

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
      }
    });
  }

  onSearchChange = (event) => {
    this.setState({ searchTerm: event.target.value });
  }

  onSearchSubmit = (event) => {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });

    if (this.needsToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm);
    }

    event.preventDefault();
  }

  setSearchTopStories = (result) => {
    const { hits, page } = result.data;
    const { searchKey, results } = this.state;

    const oldHits = results && result[searchKey]
      ? results[searchKey].hits
      : [];

    const updatedHits = [
      ...oldHits,
      ...hits,
    ];

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page },
      },
    });
  }

  render() {
    const {
      searchTerm,
      results,
      searchKey,
      error,
      fetchingStories,
    } = this.state;

    const page = (
      results
      && results[searchKey]
      && results[searchKey].page
    ) || 0;

    const list = (
      results
      && results[searchKey]
      && results[searchKey].hits
    ) || [];

    return (
      <div className="page">
        <div className="interactions">
          <Button
            onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}
            showSpinner={fetchingStories}
          >
            More
          </Button>
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          >
            Search
          </Search>
        </div>
        <div className="content">
          {error
            ? <p>Something went wrong ¯\_(ツ)_/¯</p>
            : <Table list={list} onDismiss={this.onDismiss} />
          }
        </div>
      </div>
    );
  }
}

export default App;
