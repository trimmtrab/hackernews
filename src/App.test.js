import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App from './App';
import Button from './components/Button';
import Search from './components/Search';
import Table from './components/Table'; 

Enzyme.configure({ adapter: new Adapter() });

// "describe" defines test suite
// "it" describes one test case
// in jest: it === test
describe('App', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <App />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Search', () => {

  const props = {
    value: 'Redux',
  }

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Search {...props}>Search</Search>, div);
    ReactDOM.unmountComponentAtNode(div);    
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Search {...props}>Search</Search>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Button', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <Button onClick={() => {}}>
        Give Me More
      </Button>, 
      div);
    ReactDOM.unmountComponentAtNode(div);    
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Button onClick={() => {}}>
        Give Me More
      </Button>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('onClick prop is called when clicked', () => {
    const mockOnClick = jest.fn();
    const element = shallow(
      <Button onClick={mockOnClick}>
        Mock Button
      </Button>
    );

    element.simulate('click');
    expect(mockOnClick.mock.calls.length).toBe(1);
  })
});

describe('Table', () => {

  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' },
    ],
    onDismiss: () => {},
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
    ReactDOM.unmountComponentAtNode(div);    
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Table {...props} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows two items in list', () => {
    // renders the component without its children
    const element = shallow(
      <Table {...props} />
    );
    expect(element.find('.table-row').length).toBe(2);
  })
});
