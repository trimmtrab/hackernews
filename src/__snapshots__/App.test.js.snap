// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      More
    </button>
    <form
      className="search"
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="Redux"
      />
      <button
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <div
    className="content"
  >
    <div
      className="table"
    />
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
>
  Give Me More
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form
  className="search"
>
  <input
    type="text"
  />
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a>
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
